libs/s25main/mapGenerator/RandomMap.cpp

```diff
@@ -0,0 +1,292 @@
+// Copyright (c) 2017 - 2020 Settlers Freaks (sf-team at siedler25.org)
+//
+// This file is part of Return To The Roots.
+//
+// Return To The Roots is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 2 of the License, or
+// (at your option) any later version.
+//
+// Return To The Roots is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Return To The Roots. If not, see <http://www.gnu.org/licenses/>.
+
+#include "mapGenerator/RandomMap.h"
+#include "mapGenerator/Harbors.h"
+#include "mapGenerator/HeadQuarters.h"
+#include "mapGenerator/Islands.h"
+#include "mapGenerator/Resources.h"
+#include "mapGenerator/Rivers.h"
+#include "mapGenerator/Terrain.h"
+#include "mapGenerator/TextureHelper.h"
+#include "mapGenerator/Textures.h"
+
+#include "lua/GameDataLoader.h"
+#include "libsiedler2/libsiedler2.h"
+
+#include <iostream>
+#include <stdexcept>
+
+namespace rttr { namespace mapGenerator {
+
+    unsigned GetCoastline(const MapExtent& size)
+    {
+        switch(size.x + size.y)
+        {
+            case 128:
+            case 256: return 1;
+
+            case 512:
+            case 1024: return 2;
+
+            case 2048: return 3;
+
+            default: return 4;
+        }
+    }
+
+    unsigned GetIslandRadius(const MapExtent& size)
+    {
+        switch(size.x + size.y)
+        {
+            case 128:
+            case 256: return 2;
+
+            case 512: return 3;
+
+            case 1024: return 4;
+
+            case 2048: return 5;
+
+            default: return 6;
+        }
+    }
+
+    unsigned GetSmoothRadius(const MapExtent& size)
+    {
+        switch(size.x + size.y)
+        {
+            case 128: return 2;
+
+            case 256: return 3;
+
+            case 512: return 4;
+
+            case 1024: return 5;
+
+            case 2048: return 6;
+
+            default: return 7;
+        }
+    }
+
+    unsigned GetSmoothIterations(const MapExtent& size)
+    {
+        switch(size.x + size.y)
+        {
+            case 128: return 10;
+
+            case 256: return 12;
+
+            case 512: return 15;
+
+            case 1024: return 20;
+
+            case 2048: return 25;
+
+            default: return 50;
+        }
+    }
+
+    void SmoothHeightMap(ValueMap<uint8_t>& z, const ValueRange<uint8_t>& range)
+    {
+        int radius = GetSmoothRadius(z.GetSize());
+        int iterations = GetSmoothIterations(z.GetSize());
+
+        Smooth(iterations, radius, z);
+        Scale(z, range.minimum, range.maximum);
```

Comment:
This doesn't really work 100%. Sometimes this leads to a map with not very well distributed z-values which finally results in a map without mountains. I'll have a think about a smarter approach here this weekend.

Reply:
After a bit of debugging I resolved this problem. Wasn't really related to smoothing/scaling at all.